### PR TITLE
Add support for openSUSE

### LINKED ISSUE #173: System lib search fails on openSUSE
The installer is only looking for `rpm` in the noarch site-packages, whereas on openSUSE there is only a package in the arch site-packages.

```
> rpm -ql python2-rpm
/usr/lib64/python2.7/site-packages/rpm
/usr/lib64/python2.7/site-packages/rpm-4.14.2.1-py2.7.egg-info
/usr/lib64/python2.7/site-packages/rpm/__init__.py
/usr/lib64/python2.7/site-packages/rpm/__init__.pyc
/usr/lib64/python2.7/site-packages/rpm/__init__.pyo
/usr/lib64/python2.7/site-packages/rpm/_rpm.so
/usr/lib64/python2.7/site-packages/rpm/_rpmb.so
/usr/lib64/python2.7/site-packages/rpm/_rpms.so
/usr/lib64/python2.7/site-packages/rpm/transaction.py
/usr/lib64/python2.7/site-packages/rpm/transaction.pyc
/usr/lib64/python2.7/site-packages/rpm/transaction.pyo
> python2 -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())'
/usr/lib/python2.7/site-packages
```

### PR DESCRIPTION
**Pull request by dcermak**
This is a pretty huge PR, that adds support for openSUSE.

The gist of it is:
- Add intermediate classes `NativeRpmInstaller` and `NativeRpm`, from which `Fedora{Installer|Rpm}` and `Suse{Installer|Rpm}` inherit
- fix the code in `is_fedora`, so that it does not treat openSUSE as Fedora
- move all Fedora only tests to `test_fedora_install.py` for readability purposes

I've also included some additional, but unrelated, fixes into this, e.g.: ` Mark regexes with r'' so that \d is not treated as an escape sequence` and the spelling fixes. I can remove these from this PR, so that it isn't as big.

This *should* resolve #173.

### CODE REVIEW

**Comment by junaruga** [comment]
@dcermak thanks! Amazing work!

> This is a pretty huge PR, that adds support for openSUSE.

This quite makes sense in this case, the PR is to implement SUSE feature passing the CI cases.

Let me check the PR later, after 27th or 28th August, as I am in vacation.

**Comment by dcermak** [comment]
> @dcermak thanks! Amazing work!
> 
> > This is a pretty huge PR, that adds support for openSUSE.
> 
> This "huge PR" makes sense in this case, because the PR is to implement SUSE feature passing the CI cases.
> 
> Let me check the PR later, after 27th or 28th August, as I am in vacation now.

No rush, have a nice vacation :)

**Comment by junaruga** [comment]
@dcermak I cherry-picked and merged some commits not related to SUSE features to master branch from this PR. I will comment. Because when you change your code, when the commits not related to SUSE are already merged, that makes you rebase this PR easily.

```
$ git log --pretty=oneline
70f95876a2188804d0d149e96bddffd3972bdc62 Fix typo: PRM -> RPM
  => Merged.
972bd9b624a5b36f4e4e4136cd2ca1371537d09b Fix variable spelling pacakge_* => package_*
  => Merged.
21df5d798065748f5bc10dd937f603e34c7c29d4 Add openSUSE containers to docker-compose test setup
e5f7fb50364d38c6ade508bd3513244eeef10c5b Add test for opensuse-tumbleweed /etc/os-release
4c4e023ac3d82042ff3f29284473c3d7d3eb25c7 Preliminary support for SUSE based distributions
59dec0503fa460543d1fe3383691c818069c29f0 Move all Fedora specific tests into test_install_fedora.py
  => Merged
305abe538604f9abd9f2a76587eb84ffbb1dd462 Mark regexes with r'' so that \d is not treated as an escape sequence
  => Merged.
8a4c9187627c8929ba7701a4c1cdf67d46bb08ab Move fixtures into conftest.py
  => Merged.
eb1df168fc2310064c3a1fb87db4a0ec36e57952 Add helpers and fixtures to detect SUSE based vs Fedora based
  => Partly merged.
58e67f278b2a9dfa560c8e19ec3289d0556d1aea Add intermediate class NativeRpm for Linuxes with rpm
39be2ee11342ac8645eb658ca2a81536bf4a3893 Add is_suse classmethod to Linux to detect SUSE based Linux distros
d72c7ebe0f8cfe278ff1469226b349fad0ee8bde Exclude openSUSE from being treated as Fedora based
c5753633d5eb60f37fbe8db111132377b01c880f Add intermediate class: FedoraInstaller->NativeRpmInstaller->Installer
```

* Could you rebase this PR?
* I have a question to you about `docker-compose.yml` 3 new SUSE cases "opensuse/tumbleweed", "opensuse/leap:15.1" and "opensuse/leap:15.0". Could you explain what is the difference between "tumbleweed" and "leap". Is it like Ubuntu "bionic" and "leap"? Which is newer version? "leap:15.1" and "leap:15.0" is similar with "centos7" and "centos6"? Only adding 1 leap case "opensuse/leap:15" (is alias of "15.1"?) is not enough unifying from 2 leaps cases to 1 case?

* Can you add below URLs https://hub.docker.com/r/opensuse/leap and https://hub.docker.com/r/opensuse/tumbleweed as comment in `docker-compose.yml` like other distributions?
* I like `pyNN-cov` is only for latest distribution case, `pyNN` is for the older distribution cases to save the resource (running time). And because it's hard to fix coverage testing dependency issue. Can you replace `pyNN-cov` to `pyNN` on older SUSE cases? That's the reason I asked "Which is newer version?" above.

Thanks! I think that your PR is really good!

File: docker-compose.yml [hunk 1]
```diff
@@ -121,3 +121,33 @@ services:
...
+  opensuse_leap_15_0:
+    container_name: opensuse_leap_15_0
+    build:
+      context: .
+      dockerfile: ci/Dockerfile-opensuse
+      args:
+        CONTAINER_IMAGE: opensuse/leap:15.0
+    environment:
+      TOXENV: ${TOXENV-py36-cov,py27-cov}
+    command: tox
```
Comment (junaruga) [inline docker-compose.yml]: Can you move the suse cases to above "qemu" below "ubuntu_trusty", because "qemu" is specific docker-compose task used for multi-arch container cases.

**Comment by dcermak** [comment]
"Jun Aruga" <notifications@github.com> writes:

> @dcermak I cherry-picked and merged some commits not related to SUSE features to master branch from this PR. I will comment. Because when you change your code, when the commits not related to SUSE are already merged, that makes you rebase this PR easily.
>
> ```
> $ git log --pretty=oneline
> 70f95876a2188804d0d149e96bddffd3972bdc62 Fix typo: PRM -> RPM
>   => Merged.
> 972bd9b624a5b36f4e4e4136cd2ca1371537d09b Fix variable spelling pacakge_* => package_*
>   => Merged.
> 21df5d798065748f5bc10dd937f603e34c7c29d4 Add openSUSE containers to docker-compose test setup
> e5f7fb50364d38c6ade508bd3513244eeef10c5b Add test for opensuse-tumbleweed /etc/os-release
> 4c4e023ac3d82042ff3f29284473c3d7d3eb25c7 Preliminary support for SUSE based distributions
> 59dec0503fa460543d1fe3383691c818069c29f0 Move all Fedora specific tests into test_install_fedora.py
>   => Merged
> 305abe538604f9abd9f2a76587eb84ffbb1dd462 Mark regexes with r'' so that \d is not treated as an escape sequence
>   => Merged.
> 8a4c9187627c8929ba7701a4c1cdf67d46bb08ab Move fixtures into conftest.py
>   => Merged.
> eb1df168fc2310064c3a1fb87db4a0ec36e57952 Add helpers and fixtures to detect SUSE based vs Fedora based
>   => Partly merged.
> 58e67f278b2a9dfa560c8e19ec3289d0556d1aea Add intermediate class NativeRpm for Linuxes with rpm
> 39be2ee11342ac8645eb658ca2a81536bf4a3893 Add is_suse classmethod to Linux to detect SUSE based Linux distros
> d72c7ebe0f8cfe278ff1469226b349fad0ee8bde Exclude openSUSE from being treated as Fedora based
> c5753633d5eb60f37fbe8db111132377b01c880f Add intermediate class: FedoraInstaller->NativeRpmInstaller->Installer
> ```
>
> * Could you rebase this PR?

Yes, I've done that just now.

> * I have a question to you about `docker-compose.yml` 3 new SUSE cases "opensuse/tumbleweed", "opensuse/leap:15.1" and "opensuse/leap:15.0". Could you explain what is the difference between "tumbleweed" and "leap". Is it like Ubuntu "bionic" and "leap"? Which is newer version? "leap:15.1" and "leap:15.0" is similar with "centos7" and "centos6"? Only adding 1 leap case "opensuse/leap:15" (is alias of "15.1"?) is not enough unifying from 2 leaps cases to 1 case?
>

Tumbleweed is the openSUSE rolling release/development edition and is
comparable to Fedora Rawhide. Leap on the other hand is based on SUSE
Linux Enterprise and is thus more comparable to CentOS. Leap 15.1 is
currently the newest minor release and 15.0 being the previous one
(afaik there's a new minor Leap release every year).

I've added a test for Leap 15.0 & Leap 15.1 out of habit, but
considering that 15.0 will be EOL'd this year in November, I'll just add
one Leap test for leap:15.

> * Can you add below URLs https://hub.docker.com/r/opensuse/leap and https://hub.docker.com/r/opensuse/tumbleweed as comment in `docker-compose.yml` like other distributions?

Done.

> * I like `pyNN-cov` is only for latest distribution case, `pyNN` is for the older distribution cases to save the resource (running time). And because it's hard to fix coverage testing dependency issue. Can you replace `pyNN-cov` to `pyNN` on older SUSE cases? That's the reason I asked "Which is newer version?" above.
>

In this case I'd advocate to keep the `pyNN-cov` for both Tumbleweed an
Leap, because both are pretty different from each other (like Fedora
Rawhide vs CentOS/RHEL).

> Thanks! I think that your PR is really good!
>

Glad you like it :-)

**Comment by junaruga** [comment]
Thanks for your rebasing!

> Tumbleweed is the openSUSE rolling release/development edition and is
comparable to Fedora Rawhide. Leap on the other hand is based on SUSE
Linux Enterprise and is thus more comparable to CentOS. Leap 15.1 is
currently the newest minor release and 15.0 being the previous one
(afaik there's a new minor Leap release every year).
>
> I've added a test for Leap 15.0 & Leap 15.1 out of habit, but
considering that 15.0 will be EOL'd this year in November, I'll just add
one Leap test for leap:15.

Thanks for your explanation, and unifying from Leap 15.0 & Leap 15.1 cases to "leap:15" case.

* Tumbleweed: like a Fedora Rawhide.
* Leap 15: like a CentOS 6

When Tumbleweed is like a Fedora Rawhide, what is like a Fedora 30 (Fedora stable version) in SUSE? Because when seeing "opensuse/tumbleweed", there is no version tag images. 
https://hub.docker.com/r/opensuse/tumbleweed

> In this case I'd advocate to keep the `pyNN-cov` for both Tumbleweed an
Leap, because both are pretty different from each other (like Fedora
Rawhide vs CentOS/RHEL).

Alright. Keeping `pyNN-cov` for both Tumbleweed and Leap makes sense.

Let me see the detail of your code.

File: tests/conftest.py [hunk 2]
```diff
@@ -41,15 +42,15 @@ def _get_os_id():
...
+            match = re.search(r'^ID=[\'"]?([\w_-]+)?[\'"]?$', line)
             if match:
                 os_id = match.group(1)
                 break
 
     return os_id
 
 
-_os_id = _get_os_id()
+_os_id = _get_os_id() or ()
```
Comment (junaruga) [inline tests/conftest.py]: `_get_os_id()` returns the value `ID=` as string. `()` is tuple, right?
In a case, `_os_id` becomes string, in an another case `_os_id` becomes tuple, looks not great.
Is it better to change `()` to `''` (empty string)? How do you think?

```
>>> type(())
<class 'tuple'>
>>> type('')
<class 'str'>
```

**Comment by junaruga** [comment]
I just checked the tests are not reduced like this. :) It looks good to me. 
Other parts look good to me too. Only 2 questions above.

### Current master branch (`#1037`)

https://travis-ci.org/junaruga/rpm-py-installer/builds/580331359

* fedora30:
  * py37-cov: 127 passed, 2 deselected in 137.33 seconds
  * py27-cov: 127 passed, 2 deselected, 1 warnings in 120.72 seconds
* intg:
  * 2 passed, 127 deselected in 4.46 seconds
* centos7
  * py36-cov: 127 passed, 2 deselected in 77.60 seconds
  * py34-cov: 126 passed, 1 skipped, 2 deselected in 60.81 seconds
  * py27-cov: 127 passed, 2 deselected, 1 warnings in 62.23 seconds
* centos6
  * py27: 126 passed, 1 skipped, 2 deselected, 1 warnings in 48.59 seconds
  * py26: 126 passed, 1 skipped, 2 deselected in 45.98 seconds
* ubuntu_bionic
  * py36-cov: 90 passed, 37 skipped, 2 deselected in 29.59 seconds
  * py27-cov: 90 passed, 37 skipped, 2 deselected, 1 warnings in 26.64 seconds
* ubuntu_trusty
  * py34: 89 passed, 38 skipped, 2 deselected in 16.77 seconds
  * py27: 90 passed, 37 skipped, 2 deselected, 1 warnings in 23.87 seconds

## This PR (`#1036`)

https://travis-ci.org/junaruga/rpm-py-installer/builds/579711375

* fedora30:
  * py37-cov: 136 passed, 3 skipped, 2 deselected in 133.54 seconds
  * py27-cov: 136 passed, 3 skipped, 2 deselected, 1 warnings in 92.79 seconds
* intg:
  * 2 passed, 139 deselected in 4.30 seconds
* centos7
  * py36-cov: 136 passed, 3 skipped, 2 deselected in 77.82 seconds
  * py34-cov: 135 passed, 4 skipped, 2 deselected in 58.54 seconds
  * py27-cov: 136 passed, 3 skipped, 2 deselected, 1 warnings in 61.66 seconds
* centos6
  * py27: 135 passed, 4 skipped, 2 deselected, 1 warnings in 52.57 seconds
  * py26: 135 passed, 4 skipped, 2 deselected in 43.56 seconds
* ubuntu_bionic
  * py36-cov: 101 passed, 38 skipped, 2 deselected in 27.39 seconds
  * py27-cov: 101 passed, 38 skipped, 2 deselected, 1 warnings in 25.40 seconds
* ubuntu_trusty
  * py34: 100 passed, 39 skipped, 2 deselected in 16.47 seconds
  * py27: 101 passed, 38 skipped, 2 deselected, 1 warnings in 24.50 seconds
* opensuse_tumbleweed
  * py37-cov: 104 passed, 35 skipped, 2 deselected in 31.27 seconds
  * py27-cov: 104 passed, 35 skipped, 2 deselected, 1 warnings in 30.06 seconds
* opensuse_leap_15
  * py36-cov: 104 passed, 35 skipped, 2 deselected in 33.69 seconds
  * py27-cov: 104 passed, 35 skipped, 2 deselected, 1 warnings in 30.97 seconds

**Comment by dcermak** [comment]
> When Tumbleweed is like a Fedora Rawhide, what is like a Fedora 30 (Fedora stable version) in SUSE? Because when seeing "opensuse/tumbleweed", there is no version tag images.

Unfortunately such an equivalent does not exist, in the openSUSE world you have either to use Tumbleweed and live on the bleeding edge or stick with the stable, but pretty old Leap. There is no inbetween.

File: tests/conftest.py (hunk 2 above)
Comment (dcermak) [reply tests/conftest.py]: To be honest I do not remember why I added that line. I think it was to work around some issues with CentOS, but I've rebased those commits away already.

I've pushed a fixup commit that reverts it to the original state. I'm going to rebase it away, unless it breaks the testsuite, so please don't merge this yet 😉

Comment (junaruga) [reply tests/conftest.py]: Sure. Perhaps, the issues with CentOS might be fixed by the latest commit of the master branch now. ( https://github.com/junaruga/rpm-py-installer/commit/8c8aad0669ba84255aadc13db425bec4ad241ce3 ).

When I was cherry-picking your commits, I faced CentOS6 case skipping `tests/test_install_fedora.py`. ( https://github.com/junaruga/rpm-py-installer/pull/211 )

**Comment by junaruga** [comment]
> Unfortunately such an equivalent does not exist, in the openSUSE world you have either to use Tumbleweed and live on the bleeding edge or stick with the stable, but pretty old Leap. There is no inbetween.

Alright. Thanks for the explanation.

File: tests/conftest.py (hunk 2 above)
Comment (dcermak) [reply tests/conftest.py]: > Sure. Perhaps, the issues with CentOS might be fixed by the latest commit of the master branch now. ( [8c8aad0](https://github.com/junaruga/rpm-py-installer/commit/8c8aad0669ba84255aadc13db425bec4ad241ce3) ).

No it wasn't, my check in `_is_suse` was flawed and would raise an exception on CentOS6 (because `_os_id()` returns `None` there). I've tried a different version now.

Comment (dcermak) [reply tests/conftest.py]: I've rebased the fixup and Travis should be hopefully green now.

File: install.py [hunk 3]
```diff
@@ -17,6 +17,8 @@
 from distutils.spawn import find_executable
 from distutils.sysconfig import get_python_lib
 
+import packaging.version
```
Comment (junaruga) [inline install.py]: What is `packaging`?

```
$ grep -r packaging *
install.py:import packaging.version
install.py:                most_recent_version = packaging.version.parse(
install.py:                candidate_version = packaging.version.parse(
install.py:                most_recent_release = packaging.version.parse(
install.py:                candidate_release = packaging.version.parse(
test-requirements.txt:packaging
Binary file tests/fixtures/rpm/rpm-build-libs-4.13.0.1-2.fc25.x86_64.rpm matches
Binary file tests/fixtures/rpm/rpm-build-libs-4.13.0.1-2.fc25.i686.rpm matches
Binary file tests/fixtures/rpm/rpm-devel-4.13.0.1-2.fc25.i686.rpm matches
Binary file tests/fixtures/rpm/rpm-devel-4.13.0.1-2.fc25.x86_64.rpm matches
tox-requirements.txt:packaging
```

Comment (dcermak) [reply install.py]: This module: https://github.com/pypa/packaging. It provides a convenient version parser that I'm using to get the downloaded rpms from zypper (zypper doesn't have a "download it into this directory" option, it can only download the rpm into the default cache directory, from which I am extracting the rpm).

Comment (junaruga) [reply install.py]: For example you can run `install.py` directly without using `pip` command like this.
But this command outputs the error.
Sorry I did not include this use case in the unit test.

```
$ python3 -m venv ./venv

$ source venv/bin/activate

(venv) $ python -c "$(curl -fsSL https://raw.githubusercontent.com/dcermak/rpm-py-installer/master/install.py)"
Traceback (most recent call last):
  File "<string>", line 20, in <module>
ModuleNotFoundError: No module named 'packaging'
(venv) $ 
```

Comment (dcermak) [reply install.py]: Darn, I haven't thought about that when adding the dependency on packaging.

Comment (junaruga) [reply install.py]: That we can not use pip packages as a dependency in `install.py` is a kind of current challenge for `rpm-py-installer` :<

> https://github.com/junaruga/rpm-py-installer/blob/master/docs/developers_guide.md
> The repository has below logic files. The every logic is included in the install.py. The reason is to run it directory from curl command without pip. I think that is useful to debug before actual release. But it makes the source code hard to read. If we can compress the logic from .py files to one py file including zipped binary code like get-pip.py, we can split the install.py to several py files.

Is it possible to do "get the downloaded rpms from zypper" without using `packaging`? `install.py` has `Cmd.curl_remote_name` class method to download a file from web like curl command.

File: tests/conftest.py (hunk 2 above)
Comment (junaruga) [reply tests/conftest.py]: Sure, and Travis is green now :)

File: install.py (hunk 3 above)
Comment (dcermak) [reply install.py]: > Is it possible to do "get the downloaded rpms from zypper" without using `packaging`?

Of course, I can implement the version parsing and comparison myself. It's just more work.

Comment (dcermak) [reply install.py]: I went a little spelunking into the code base and found that you actually already implemented version parsing (albeit simpler but good enough for RPM version & releases). Let's see whether Travis is happy too, as it works on my machine™

Comment (junaruga) [reply install.py]: Oh that is `Utils.version_str2tuple`. I forget the existence. Travis is happy too!

Comment (junaruga) [reply install.py]: I tried to run above command to install rpm-py-installer directly without pip.

```
$ python --version
Python 3.7.4

$ source venv/bin/activate
(venv) $ python -c "$(curl -fsSL https://raw.githubusercontent.com/dcermak/rpm-py-installer/master/install.py)"
[INFO] Installing...
[INFO] Created working directory '/tmp/tmpez5bt5dr-rpm-py-installer'
[INFO] Downloading archive. 'https://github.com/rpm-software-management/rpm/archive/rpm-4.14.2.1-release.tar.gz'.
[INFO] Removed working directory '/tmp/tmpez5bt5dr-rpm-py-installer'
[INFO] Done successfully.
(venv) $ Traceback (most recent call last):
```

Then I got a unknown message "Traceback (most recent call last):".
Do you see it on your machine?

```
(venv) $ pip list
Package    Version 
---------- --------
pip        19.0.3  
rpm        4.14.2.1
setuptools 40.8.0  
You are using pip version 19.0.3, however version 19.2.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

Comment (junaruga) [reply install.py]: The message "Traceback (most recent call last):" is not outputted for `install.py` on this repository's latest master branch.

```
(venv_current_master) $ python -c "$(curl -fsSL https://raw.githubusercontent.com/junaruga/rpm-py-installer/master/install.py)"
[INFO] Installing...
[INFO] Created working directory '/tmp/tmp4_fsh_5g-rpm-py-installer'
[INFO] Downloading archive. 'https://github.com/rpm-software-management/rpm/archive/rpm-4.14.2.1-release.tar.gz'.
[INFO] Removed working directory '/tmp/tmp4_fsh_5g-rpm-py-installer'
[INFO] Done successfully.
(venv_current_master) $ 
```

File: install.py [hunk 4]
```diff
@@ -1962,10 +2131,30 @@ def convert_to_int(string):
                 value = string
             return value
 
-        version_info_list = [convert_to_int(s) for s in version_info_list]
+        version_info_list = (convert_to_int(s) for s in version_info_list)
 
         return tuple(version_info_list)
 
+    @staticmethod
+    def _version_cmp(operation):
```
Comment (junaruga) [inline install.py]: `def _version_cmp` has `_method`. You wanted to implement a kind of private method, right? Or you did put the heading the `_` by typo?
If it means a private method, could you move it to after a public method `version_equal`? This is like `class Python` in `install.py`.

File: install.py [hunk 5]
```diff
@@ -1944,14 +2113,14 @@ def mkdir_p(cls, path):
 class Utils(object):
     """A general utility class."""
 
-    @classmethod
-    def version_str2tuple(cls, version_str):
-        """Version info.
+    @staticmethod
```
Comment (junaruga) [inline install.py]: Is `@staticmethod` better than `@classmethod`? Why do you prefer it?
I am not familiar with a recommended Python style. I just found below article about it.
https://stackoverflow.com/questions/136097/what-is-the-difference-between-staticmethod-and-classmethod#1669524

File: install.py (hunk 4 above)
Comment (dcermak) [reply install.py]: Yes, the `_` prefix is intentional. I'll move it after the public methods.

File: install.py (hunk 5 above)
Comment (dcermak) [reply install.py]: Not necessarily "better", I just find it more suitable here. `@classmethod` is used to mark methods which just need access to the class' type but not to the specific instance. But this method doesn't even require that, in fact the `Utils` class is "just a namespace" (because everything has to be in a single file). So using `@staticmethod` appeared to me better here, as these are just free functions that are grouped inside a class.

File: install.py (hunk 3 above)
Comment (dcermak) [reply install.py]: Hm, unfortunately I cannot reproduce this additional `Traceback (most recent call last):` output. In which environment did you run it? I've tried a fresh Tumbleweed, Leap 15.1 and Fedora 30 Vagrant box.

File: install.py (hunk 5 above)
Comment (junaruga) [reply install.py]: Alright. I respect your opinion. :)

File: install.py (hunk 4 above)
Comment (junaruga) [reply install.py]: Sure, thanks for moving it.

File: install.py (hunk 3 above)
Comment (junaruga) [reply install.py]: My environment where I faced the issue is, my local machine. Fedora 30.
The python is like this.
But surprisingly when I tried it again today, it does not show the `Traceback ...` message. It is no problem. :)

```
$ which python3
/usr/bin/python3

$ rpm -qf /usr/bin/python3
python3-3.7.4-1.fc30.x86_64

$ /usr/bin/python3 --version
Python 3.7.4

$ python3 -m venv ./ven

$ source venv/bin/activate

(venv) $ pip list
Package    Version
---------- -------
pip        19.0.3 
setuptools 40.8.0 
You are using pip version 19.0.3, however version 19.2.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

(venv) $ python -c "$(curl -fsSL https://raw.githubusercontent.com/dcermak/rpm-py-installer/master/install.py)"
[INFO] Installing...
[INFO] Created working directory '/tmp/tmpe1ciftxx-rpm-py-installer'
[INFO] Downloading archive. 'https://github.com/rpm-software-management/rpm/archive/rpm-4.14.2.1-release.tar.gz'.
[INFO] Removed working directory '/tmp/tmpe1ciftxx-rpm-py-installer'
[INFO] Done successfully.

(venv) $ pip list
Package    Version 
---------- --------
pip        19.0.3  
rpm        4.14.2.1
setuptools 40.8.0  
You are using pip version 19.0.3, however version 19.2.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

(venv) $ python -c "import rpm; print(rpm.__version__)"
4.14.2.1
```

**Comment by junaruga** [comment]
@dcermak I think now we are ready to merge this PR. I like to merge this PR with "Squash and merge" option unifying the 8 commits to 1 commit. Because I want to guarantee the CI is passed for each commit on master branch. Do you like squashing the 8 commits by yourself? If you do not mind about it, I just do "squash and merge" this PR from this page, keeping each commit's your comment.

**Comment by dcermak** [comment]
@junaruga I've squashed the commits and force pushed them. Let's see whether Travis is still happy.

**Comment by junaruga** [comment]
Travis is happy!

**Comment by junaruga** [comment]
Are you ready to merge it now?

**Comment by dcermak** [comment]
Yes, please merge it at your convenience.

**Comment by junaruga** [comment]
Now it was merged to master branch.
@dcermak thank you for your great work. I appreciate it.

My plan is that next is to update document, at least README - Supported environments, and other parts if it's necessary. Then release new version.
https://github.com/junaruga/rpm-py-installer/blob/master/README.md#supported-environments

Do you like to send a pull-request for that?

**Comment by junaruga** [comment]
You can use `[skip *]` keyword in the commit message, if you send pull-request that has only document's modifications.
https://docs.travis-ci.com/user/customizing-the-build/#skipping-a-build

**Comment by jayvdb** [comment]
@dcermak Thanks so much for this!  After it is released, do you plan on creating/submitting the openSUSE package?

**Comment by dcermak** [comment]
@junaruga I'll send a PR to update the docs.

@jayvdb This package shouldn't be packaged as an RPM at all: it's only purpose is to install the python rpm bindings for packages obtained via pip. If you want to do the same for a rpm package, then simply add a `BuildRequires: python3-rpm` and patch out the dependency on `rpm-py-installer`.

**Comment by jayvdb** [comment]
@dcermak ,

1. patching out the dependency on `rpm-py-installer` shouldnt be necessary, and is an approach which does not scale.  rpm is not special.  There are many packages like this one which are 'check-system-is-ok-or-install-system-stuff'.  They are becoming more common, and they are being added as dependencies to other projects.
2. While the package runtime code is empty, the metadata in the egg-info is not -- its presence is mandatory for pip to know that it is already installed, so that SuSE users can install the rpm and then when the user runs pip it will skip fetching and "installing" this package; e.g. in `tox --sitepackages`
3. SuSE needs to know this package always works on the OS they deliver -- the way to do that is to have it as a package in OBS running against whatever targets are desirable.  This is quite important given the openSUSE layout is significantly different, and because SuSE has arch which are not usually tested by "developer CI".

**Comment by dcermak** [comment]
John Vandenberg <notifications@github.com> writes:

> @dcermak ,
>
> 1. patching out the dependency on `rpm-py-installer` shouldnt be necessary, and is an approach which does not scale.  rpm is not special.  There are many packages like this one which are 'check-system-is-ok-or-install-system-stuff'.  They are becoming more common, and they are being added as dependencies to other projects.

If you have a package that *requires* rpm-py-installer at all times,
then that is a bug in that package. See for instance Q6 in the FAQ
(user_guide.md) or #134.

tl;dr; A package should actually first check whether something already
provides `rpm` and `rpm-python` before requiring `rpm-py-installer`.

> 2. While the package runtime code is empty, the metadata in the egg-info is not -- its presence is mandatory for pip to know that it is already installed, so that SuSE users can install the rpm and then when the user runs pip it will skip fetching and "installing" this package; e.g. in `tox --sitepackages`

This would be also solved by depending packages simply checking whether
something provides `rpm` & `rpm-python` instead of forcing the
installation of `rpm-py-installer`.

> 3. SuSE needs to know this package always works on the OS they deliver -- the way to do that is to have it as a package in OBS running against whatever targets are desirable.  This is quite important given the openSUSE layout is significantly different, and because SuSE has arch which are not usually tested by "developer CI".

Yes. But an rpm of rpm-py-installer is *not* going to solve this at
all. Since this package has no runtime component, you'll just have an
empty rpm that only `Requires: rpm` and `Provides: python-rpm-py-installer`.
You won't actually test the installer at all, because you cannot and do
not want to run it in rpmbuild.

A (imho) good enough solution is the CI employed by this project: it
runs all the tests on Tumbleweed and Leap 15. There's of course always
room for improvements, but in this case OBS is definitely not the
answer.

**Comment by jayvdb** [comment]
> If you have a package that *requires* rpm-py-installer at all times, then that is a bug in that package. See for instance Q6 in the FAQ (user_guide.md) or #134. tl;dr; A package should actually first check whether something already provides `rpm` and `rpm-python` before requiring `rpm-py-installer`.

That is not possible if the project wants to distribute wheels - its dependencies are metadata and can not be determined at runtime - there is no setup.py.

This project, rpm-py-installer, can not be a wheel for obvious reasons, and so other projects defer to this project to handle the problem for them.

As a result, instead of all rpm-related projects being unable to distribute wheels, only this project is without wheels.

See https://github.com/junaruga/rpm-py-installer/network/dependents?package_id=UGFja2FnZS01MjU1NDQxNw%3D%3D for two uses GitHub knows about.

[koji](https://github.com/hroncok/koji/blob/tilde/setup.py) is another one.

> or #134

Yes, this is about https://gitlab.com/rpkg/rpkg2/blame/master/requirements/pypi.txt which has dropped the dependency.

The choice by RedHat to not package `rpm-py-installer` is a reasonable approach when there is only one or two projects using it, but that number is going up.

Anyways, you obviously wont be packaging it.  If I ever do get back to whatever it was where I wanted/needed it, I will ping you if I submit a packaging request so you we can continue there.

**Comment by dcermak** [comment]
"John Vandenberg" <notifications@github.com> writes:

>> If you have a package that *requires* rpm-py-installer at all times, then that is a bug in that package. See for instance Q6 in the FAQ (user_guide.md) or #134. tl;dr; A package should actually first check whether something already provides `rpm` and `rpm-python` before requiring `rpm-py-installer`.
>
> That is not possible if the project wants to distribute wheels - its dependencies are metadata and can not be determined at runtime - there is no setup.py.
>

Hm, that is indeed a problem, that I did not consider.

> This project, rpm-py-installer, can not be a wheel for obvious reasons, and so other projects defer to this project to handle the problem for them.
>
> As a result, instead of all rpm-related projects being unable to distribute wheels, only this project is without wheels.
>
> See https://github.com/junaruga/rpm-py-installer/network/dependents?package_id=UGFja2FnZS01MjU1NDQxNw%3D%3D for two uses GitHub knows about.
>
> [koji](https://github.com/hroncok/koji/blob/tilde/setup.py) is another one.
>
>> or #134
>
> Yes, this is about https://gitlab.com/rpkg/rpkg2/blame/master/requirements/pypi.txt which has dropped the dependency.
>
> The choice by RedHat to not package `rpm-py-installer` is a reasonable approach when there is only one or two projects using it, but that number is going up.
>

Patching just two projects would be ok by me, as that task is relatively
trivial (a sed call to setup.py/requirements.txt and to
*.egg-info/requires.txt). However, if a lot of packages start to depend
on rpm-py-installer, then we might have to reconsider that approach.

Ideally this should be solved by rpm upstream.

> Anyways, you obviously wont be packaging it.

That is correct, because it will get rejected and I have no need for
this package as an rpm.

> If I ever do get back to whatever it was where I wanted/needed it, I will ping you if I submit a packaging request so you we can continue there.
>

Sure, my OBS username is dancermak. However, bear in mind that I am not
a reviewer/maintainer of devel:languages:python. So if they say no: I
cannot change that.
